### ISSUE TITLE
WASM Compile Target?

### ISSUE THREAD

**Issue by darth-cheney** [OpenSmalltalk/opensmalltalk-vm#387]
Hello all! As some of you might be aware, there has been a lot of progress in the web-world in developing [WebAssembly](https://webassembly.org/). Many environments and languages can now compile to the wasm binary format. Projects are doing things like [putting full Python interpreters and environments in the browser](https://github.com/iodide-project/pyodide), etc. There is even a nascent movement to come up with a consistent [system interface standard](https://hacks.mozilla.org/2019/03/standardizing-wasi-a-webassembly-system-interface/) for WebAssembly, allowing developers to run wasm code on any platform -- not just the web.
  
My question is this: what would it take to get the opensmalltalk-vm to compile to WebAssembly and has anyone involved with it considered the option so far?
  
There would be a great many advantages to running a Smalltalk in the browser this way.

**Comment by johnmci** [comment]
$10,000,000 usa

**Comment by ccrraaiigg** [comment]
Yes, I've been following WASM since it attained full browser coverage in 2017. I've built the OpenSmalltalk stack interpreter with emscripten, no major gotchas. Of course we want a WASM Cog. We're waiting for WASM to do garbage collection, and provide an API for running your own generated native code. I expect both of these to arrive in the next two years.

In the meantime, I'm getting very good results running Squeak in web browser and on Node with SqueakJS. The simple bytecode-to-JS dynamic translator that Bert wrote for it is a big win. You get a decent translation of bytecodes to native code via the underlying JS engine (V8 et al).

**Comment by ccrraaiigg** [comment]
It takes a Smalltalk CompiledMethod object, and transcribes the instructions (bytecodes) into a JavaScript method, and makes sure to run that instead the next time the CompiledMethod would normally run. Then V8 optimizes the hell out of that transcribed JS method. It's pretty neat, and makes a huge difference for lots of things, including the UI.
